src/map/battle_skill_factory.cpp

```diff
@@ -0,0 +1,72 @@
+#include "battle_skill_factory.hpp"
+#include "battle_skill_categories.hpp"
+#include "battle_skills.hpp"
+#include "skill.hpp"
+
+BattleSkillFactory::BattleSkillFactory()
+{
+    register_all_skills();
+}
+
+BattleSkillFactory &BattleSkillFactory::instance()
+{
+    static BattleSkillFactory instance;
+    return instance;
+}
+
+std::shared_ptr<BattleSkill> BattleSkillFactory::get_skill(uint16 skill_id) const
+{
+    auto it = skill_db.find(skill_id);
+    if (it != skill_db.end())
+    {
+        return it->second;
+    }
+
+    return nullptr;
+}
+
+bool BattleSkillFactory::has_skill(uint16 skill_id) const
+{
+    return skill_db.find(skill_id) != skill_db.end();
+}
+
+void BattleSkillFactory::register_all_skills()
+{
+    register_weapon_skills();
+    register_magic_skills();
+    register_misc_skills();
+    register_passive_skills();
+}
+
+void BattleSkillFactory::register_weapon_skills()
+{
+    register_skills<BashSkill>({SM_BASH, MS_BASH});
```

Comment:
Personally I don't like this, I think SM_BASH and MS_BASH should be two different classes
You could make BashSkill and MercenaryBashSkill inherit BashSkill